docs/code-generation.md

```diff
@@ -0,0 +1,135 @@
+# Code generation
+
+In order to keep the code for all the service controllers consistent, we will
+use a strategy of generating the custom resource definitions and controller
+code stubs for new AWS services.
+
+To generate the CRDs and controller stub code, we investigated a number of
+options:
+
+* home-grown custom code generator
+* [kudo](https://kudo.dev)
+* [kubebuilder](github.com/kubernetes-sigs/kubebuilder)
+* a hybrid custom code generator + sigs.kubernetes.io/controller-tools (CR)
+
+The original AWS Service Operator used a [custom-built generator][1] that
+processed [YAML files][2] describing the AWS service and used [templates][3] to
+[generate CRDs][4], the [controller code][5] itself and the [Go types][6] that
+mirror the CRDs. It's worth noting that the CRDs *and* the controller code that
+was generated by the original ASO was very tightly coupled to CloudFormation.
+In fact, the CRDs for individual AWS services like S3 or RDS were thin wrappers
+around CloudFormation stacks that described the object being operated upon.
+
+[1]: https://github.com/amazon-archives/aws-service-operator/tree/master/code-generation
+[2]: https://github.com/amazon-archives/aws-service-operator/tree/master/models
+[3]: https://github.com/amazon-archives/aws-service-operator/tree/master/code-generation/pkg/codegen/assets
+[4]: https://github.com/amazon-archives/aws-service-operator/blob/b4befd62322a57ac78aa39ea08771fc32912592a/code-generation/pkg/codegen/assets/aws-service-operator.yaml.templ#L13-L31
+[5]: https://github.com/amazon-archives/aws-service-operator/blob/master/code-generation/pkg/codegen/assets/operator.go.templ
+[6]: https://github.com/amazon-archives/aws-service-operator/blob/master/code-generation/pkg/codegen/assets/types.go.templ
+
+kudo is a platform for building Kubernetes Operators. It stores state in its
+own kudo.dev CRDs and allows users to define "plans" for a deployed application
+to deploy itself. We determined that kudo was not a particularly good fit for
+ASO for a couple reasons. First, we needed a way to generate CRDs in several
+API groups (s3.aws.com and iam.aws.com for example) and the ASO controller code
+isn't deploying an "application" that needs to have a controlled deployment
+plan. Instead, ASO is a controller that facilitates creation and management of
+various AWS service objects using Kubernetes CRD instances.
+
+kubebuilder is the the recommended upstream tool for generating CRDs and
+controller stub code. It is a Golang binary that creates the scaffolding for
```

Comment:
Nit: While `Golang` is not incorrect, I'd prefer the simpler, short, and in our context unambiguous `Go`.

Reply:
done :)